src/tools/w3dviewer/w3dviewdoc.cpp

```diff
@@ -467,7 +467,7 @@ void CW3DViewDoc::SetRenderObject(RenderObjClass *robj, bool useRegularCameraRes
     }
 
     robj->Set_Animation();
-    m_model = robj;

```

Comment:
```
==20628==ERROR: AddressSanitizer: heap-use-after-free on address 0x45fbdf48 at pc 0x000c52be bp 0x00aff060 sp 0x00aff054
READ of size 4 at 0x45fbdf48 thread T0
    #0 0xc52bd in RefCountClass::Release_Ref D:\Projects\TheAssemblyArmada\Thyme\src\w3d\lib\refcount.h:46
    #1 0xc4dd7 in Ref_Ptr_Release<RenderObjClass> D:\Projects\TheAssemblyArmada\Thyme\src\w3d\lib\refcount.h:81
    #2 0x1147fe in CW3DViewDoc::SetRenderObject D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\w3dviewdoc.cpp:419
    #3 0xd7d9f in CDataTreeView::Select D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\datatreeview.cpp:365
    #4 0xd74bc in CDataTreeView::OnSelectTree D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\datatreeview.cpp:96
    #5 0x5935fc15  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x101efc15)
    #6 0x593fb768  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1028b768)
    #7 0x593fb55b  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1028b55b)
    #8 0x593fb60d  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1028b60d)
    #9 0x593fa9ec  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1028a9ec)
    #10 0x59414795  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x102a4795)
    #11 0x593f9fd1  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10289fd1)
    #12 0x593f9ecc  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10289ecc)
    #13 0x593f8183  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10288183)
    #14 0x593f83d3  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x102883d3)
    #15 0x592d7653  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10167653)
    #16 0x763716ea in AddClipboardFormatListener+0x4a (C:\WINDOWS\System32\USER32.dll+0x69e416ea)
    #17 0x76367b39 in GetClassLongW+0x7a9 (C:\WINDOWS\System32\USER32.dll+0x69e37b39)
    #18 0x76368a58 in GetParent+0x928 (C:\WINDOWS\System32\USER32.dll+0x69e38a58)
    #19 0x76365e3c in SendMessageW+0x4c (C:\WINDOWS\System32\USER32.dll+0x69e35e3c)
    #20 0x6f32f090 in Ordinal342+0x8e0 (C:\WINDOWS\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_5.82.19041.4355_none_c0dc01d438beab35\COMCTL32.dll+0x5bf9f090)
    #21 0x6f3585ed in TrackMouseEvent+0x2664d (C:\WINDOWS\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_5.82.19041.4355_none_c0dc01d438beab35\COMCTL32.dll+0x5bfc85ed)
    #22 0x6f358f29 in TrackMouseEvent+0x26f89 (C:\WINDOWS\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_5.82.19041.4355_none_c0dc01d438beab35\COMCTL32.dll+0x5bfc8f29)
    #23 0x6f35a1ca in TrackMouseEvent+0x2822a (C:\WINDOWS\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_5.82.19041.4355_none_c0dc01d438beab35\COMCTL32.dll+0x5bfca1ca)
    #24 0x6f35b5bf in TrackMouseEvent+0x2961f (C:\WINDOWS\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_5.82.19041.4355_none_c0dc01d438beab35\COMCTL32.dll+0x5bfcb5bf)
    #25 0x763716ea in AddClipboardFormatListener+0x4a (C:\WINDOWS\System32\USER32.dll+0x69e416ea)
    #26 0x76367b39 in GetClassLongW+0x7a9 (C:\WINDOWS\System32\USER32.dll+0x69e37b39)
    #27 0x763672f5 in GetSystemMetricsForDpi+0xc45 (C:\WINDOWS\System32\USER32.dll+0x69e372f5)
    #28 0x763543ca in CallWindowProcA+0x1a (C:\WINDOWS\System32\USER32.dll+0x69e243ca)
    #29 0x593f8e23  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10288e23)
    #30 0x593f9eed  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10289eed)
    #31 0x593f8183  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10288183)
    #32 0x593f83d3  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x102883d3)
    #33 0x592d7653  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10167653)
    #34 0x763716ea in AddClipboardFormatListener+0x4a (C:\WINDOWS\System32\USER32.dll+0x69e416ea)
    #35 0x76367b39 in GetClassLongW+0x7a9 (C:\WINDOWS\System32\USER32.dll+0x69e37b39)
    #36 0x76366470 in DispatchMessageW+0x4f0 (C:\WINDOWS\System32\USER32.dll+0x69e36470)
    #37 0x763580ef in DispatchMessageA+0xf (C:\WINDOWS\System32\USER32.dll+0x69e280ef)
    #38 0x593e50fd  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x102750fd)
    #39 0x5940e243  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1029e243)
    #40 0x28543a in __scrt_common_main_seh D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
    #41 0x7696fcc8 in BaseThreadInitThunk+0x18 (C:\WINDOWS\System32\KERNEL32.DLL+0x6b81fcc8)
    #42 0x773f80cd in RtlGetAppContainerNamedObjectPath+0x11d (C:\WINDOWS\SYSTEM32\ntdll.dll+0x4b2e80cd)
    #43 0x773f809d in RtlGetAppContainerNamedObjectPath+0xed (C:\WINDOWS\SYSTEM32\ntdll.dll+0x4b2e809d)

0x45fbdf48 is located 8 bytes inside of 316-byte region [0x45fbdf40,0x45fbe07c)
freed by thread T0 here:
    #0 0x5965947d in __asan_wrap_RtlValidateHeap+0x196d (D:\Projects\TheAssemblyArmada\Thyme\.solution\vs2022\RelWithDebInfo\clang_rt.asan_dynamic-i386.dll+0x1003947d)
    #1 0x5966a41c in _asan_wrap__except_handler4+0x47ec (D:\Projects\TheAssemblyArmada\Thyme\.solution\vs2022\RelWithDebInfo\clang_rt.asan_dynamic-i386.dll+0x1004a41c)
    #2 0x284683 in operator delete D:\a\_work\1\s\src\vctools\asan\llvm\compiler-rt\lib\asan\asan_win_delete_scalar_size_thunk.cpp:44
    #3 0x16f85b in HLodClass::`scalar deleting destructor'+0x1b (D:\Projects\TheAssemblyArmada\Thyme\.solution\vs2022\RelWithDebInfo\w3dview.exe+0x50f85b)
    #4 0x10e396 in RefCountClass::Delete_This D:\Projects\TheAssemblyArmada\Thyme\src\w3d\lib\refcount.h:53
    #5 0xc5340 in RefCountClass::Release_Ref D:\Projects\TheAssemblyArmada\Thyme\src\w3d\lib\refcount.h:49
    #6 0x167e36 in SimpleSceneClass::Remove_Render_Object D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\scene.cpp:103
    #7 0x1585ef in RenderObjClass::Remove D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\rendobj.cpp:332
    #8 0x112644 in CW3DViewDoc::RemoveRenderObject D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\w3dviewdoc.cpp:214
    #9 0x1147f2 in CW3DViewDoc::SetRenderObject D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\w3dviewdoc.cpp:418
    #10 0xd7d9f in CDataTreeView::Select D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\datatreeview.cpp:365
    #11 0xd74bc in CDataTreeView::OnSelectTree D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\datatreeview.cpp:96
    #12 0x5935fc15  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x101efc15)
    #13 0x593fb768  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1028b768)
    #14 0x593fb55b  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1028b55b)
    #15 0x593fb60d  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1028b60d)
    #16 0x593fa9ec  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1028a9ec)
    #17 0x59414795  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x102a4795)
    #18 0x593f9fd1  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10289fd1)
    #19 0x593f9ecc  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10289ecc)
    #20 0x593f8183  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10288183)
    #21 0x593f83d3  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x102883d3)
    #22 0x592d7653  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10167653)
    #23 0x763716ea in AddClipboardFormatListener+0x4a (C:\WINDOWS\System32\USER32.dll+0x69e416ea)
    #24 0x76367b39 in GetClassLongW+0x7a9 (C:\WINDOWS\System32\USER32.dll+0x69e37b39)
    #25 0x76368a58 in GetParent+0x928 (C:\WINDOWS\System32\USER32.dll+0x69e38a58)
    #26 0x76365e3c in SendMessageW+0x4c (C:\WINDOWS\System32\USER32.dll+0x69e35e3c)
    #27 0x6f32f090 in Ordinal342+0x8e0 (C:\WINDOWS\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_5.82.19041.4355_none_c0dc01d438beab35\COMCTL32.dll+0x5bf9f090)

previously allocated by thread T0 here:
    #0 0x5965957d in __asan_wrap_RtlValidateHeap+0x1a6d (D:\Projects\TheAssemblyArmada\Thyme\.solution\vs2022\RelWithDebInfo\clang_rt.asan_dynamic-i386.dll+0x1003957d)
    #1 0x591d9035  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10069035)
    #2 0x171712 in HLodPrototypeClass::Create D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\hlod.cpp:382
    #3 0x11a209 in W3DAssetManager::Create_Render_Obj D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\assetmgr.cpp:280
    #4 0xd680b in CDataTreeView::GetRenderObj D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\datatreeview.cpp:427
    #5 0xd7d8d in CDataTreeView::Select D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\datatreeview.cpp:364
    #6 0xd74bc in CDataTreeView::OnSelectTree D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\datatreeview.cpp:96
    #7 0x5935fc15  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x101efc15)
    #8 0x593fb768  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1028b768)
    #9 0x593fb55b  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1028b55b)
    #10 0x593fb60d  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1028b60d)
    #11 0x593fa9ec  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1028a9ec)
    #12 0x59414795  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x102a4795)
    #13 0x593f9fd1  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10289fd1)
    #14 0x593f9ecc  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10289ecc)
    #15 0x593f8183  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10288183)
    #16 0x593f83d3  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x102883d3)
    #17 0x592d7653  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10167653)
    #18 0x763716ea in AddClipboardFormatListener+0x4a (C:\WINDOWS\System32\USER32.dll+0x69e416ea)
    #19 0x76367b39 in GetClassLongW+0x7a9 (C:\WINDOWS\System32\USER32.dll+0x69e37b39)
    #20 0x76368a58 in GetParent+0x928 (C:\WINDOWS\System32\USER32.dll+0x69e38a58)
    #21 0x76365e3c in SendMessageW+0x4c (C:\WINDOWS\System32\USER32.dll+0x69e35e3c)
    #22 0x6f32f090 in Ordinal342+0x8e0 (C:\WINDOWS\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_5.82.19041.4355_none_c0dc01d438beab35\COMCTL32.dll+0x5bf9f090)
    #23 0x6f3585ed in TrackMouseEvent+0x2664d (C:\WINDOWS\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_5.82.19041.4355_none_c0dc01d438beab35\COMCTL32.dll+0x5bfc85ed)
    #24 0x6f358f29 in TrackMouseEvent+0x26f89 (C:\WINDOWS\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_5.82.19041.4355_none_c0dc01d438beab35\COMCTL32.dll+0x5bfc8f29)
    #25 0x6f35a1ca in TrackMouseEvent+0x2822a (C:\WINDOWS\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_5.82.19041.4355_none_c0dc01d438beab35\COMCTL32.dll+0x5bfca1ca)
    #26 0x6f35b5bf in TrackMouseEvent+0x2961f (C:\WINDOWS\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_5.82.19041.4355_none_c0dc01d438beab35\COMCTL32.dll+0x5bfcb5bf)
    #27 0x763716ea in AddClipboardFormatListener+0x4a (C:\WINDOWS\System32\USER32.dll+0x69e416ea)

SUMMARY: AddressSanitizer: heap-use-after-free D:\Projects\TheAssemblyArmada\Thyme\src\w3d\lib\refcount.h:46 in RefCountClass::Release_Ref
Shadow bytes around the buggy address:
  0x38bf7b90: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x38bf7ba0: fd fd fd fd fd fd fd fd fd fa fa fa fa fa fa fa
  0x38bf7bb0: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x38bf7bc0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x38bf7bd0: fd fd fd fd fd fd fd fd fd fd fa fa fa fa fa fa
=>0x38bf7be0: fa fa fa fa fa fa fa fa fd[fd]fd fd fd fd fd fd
  0x38bf7bf0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x38bf7c00: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x38bf7c10: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x38bf7c20: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x38bf7c30: fd fd fd fd fd fd fd fd fd fd fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
Address Sanitizer Error: Use of deallocated memory
```

Reply:
This fix isn't correct, correct fix is an Add_Ref call immediately after the = line.

Reply:
Ok. Does same thing, but will need an assert then.